internal/controllers/machine/machine_controller_noderef.go

```diff
@@ -269,16 +240,48 @@ func (r *Reconciler) getNode(ctx context.Context, c client.Reader, providerID *n
 	return &nodeList.Items[0], nil
 }
 
-func (r *Reconciler) reconcileNodeTaints(ctx context.Context, remoteClient client.Client, node *corev1.Node) error {
-	patchHelper, err := patch.NewHelper(node, remoteClient)
-	if err != nil {
-		return errors.Wrapf(err, "failed to create patch helper for Node %s", klog.KObj(node))
+// PatchNode is required to workaround an issue on Node.Status.Address which is incorrectly annotated as patchStrategy=merge
+// and this causes SSA patch to fail in case there are two address with the same key https://github.com/kubernetes-sigs/cluster-api/issues/8417
+func (r *Reconciler) patchNode(ctx context.Context, node *corev1.Node, newLabels, newAnnotations map[string]string) error {
+	newNode := node.DeepCopy()
+
+	// Adds all the annotation CAPI sets on the node.
+	hasAnnotationChanges := annotations.AddAnnotations(newNode, newAnnotations)
+
+	// Adds all the labels from the machine.
+	// NOTE: in order to handle deletion we are tracking the labels set from machine in an annotation.
+	// at the next reconcile we are going to use this for deleting labels previously set machine, but
+	// not present anymore. Labels not set from machines should be always preserved.
+	if newNode.Labels == nil {
+		newNode.Labels = make(map[string]string)
+	}
+	hasLabelChanges := false
+	labelsFromPreviousReconcile := strings.Split(newNode.Annotations[clusterv1.LabelsFromMachineAnnotation], ",")
+	if len(labelsFromPreviousReconcile) == 1 && labelsFromPreviousReconcile[0] == "" {
+		labelsFromPreviousReconcile = []string{}
+	}
+	labelsFromCurrentReconcile := []string{}
+	for k, v := range newLabels {
+		if cur, ok := newNode.Labels[k]; !ok || cur != v {
+			newNode.Labels[k] = v
+			hasLabelChanges = true
+		}
+		labelsFromCurrentReconcile = append(labelsFromCurrentReconcile, k)
 	}
-	// Drop the NodeUninitializedTaint taint on the node.
-	if taints.RemoveNodeTaint(node, clusterv1.NodeUninitializedTaint) {
-		if err := patchHelper.Patch(ctx, node); err != nil {
-			return errors.Wrapf(err, "failed to patch Node %s to modify taints", klog.KObj(node))
+	for _, k := range labelsFromPreviousReconcile {
+		if _, ok := newLabels[k]; !ok {
+			delete(newNode.Labels, k)
+			hasLabelChanges = true
 		}
 	}
-	return nil
+	annotations.AddAnnotations(newNode, map[string]string{clusterv1.LabelsFromMachineAnnotation: strings.Join(labelsFromCurrentReconcile, ",")})
+
+	// Drop the NodeUninitializedTaint taint on the node given that we are reconciling labels.
+	hasTaintChanges := taints.RemoveNodeTaint(newNode, clusterv1.NodeUninitializedTaint)
+
+	if !hasAnnotationChanges && !hasLabelChanges && !hasTaintChanges {
+		return nil
+	}
+
+	return r.Client.Patch(ctx, newNode, client.StrategicMergeFrom(node))
```

Comment:
We should use the `remoteClient` here. `r.Client` point to the management cluster,  we need to use the client that woks on the workload cluster.